File: clevertap-core/src/main/java/com/clevertap/android/sdk/db/DBManager.java
package com.clevertap.android.sdk.db;

import android.content.Context;
import android.content.SharedPreferences;
import androidx.annotation.WorkerThread;
import com.clevertap.android.sdk.CTLockManager;
import com.clevertap.android.sdk.CleverTapInstanceConfig;
import com.clevertap.android.sdk.Constants;
import com.clevertap.android.sdk.StorageHelper;
import com.clevertap.android.sdk.events.EventGroup;
import java.util.Iterator;
import org.json.JSONException;
import org.json.JSONObject;

public class DBManager extends BaseDatabaseManager {

    private DBAdapter dbAdapter;

    private final CTLockManager ctLockManager;

    private final CleverTapInstanceConfig config;

    public DBManager(CleverTapInstanceConfig config,
            CTLockManager ctLockManager) {
        this.config = config;
        this.ctLockManager = ctLockManager;
    }

    @WorkerThread
    @Override
    public synchronized DBAdapter loadDBAdapter(final Context context) {
        if (dbAdapter == null) {
            dbAdapter = new DBAdapter(context, config);
            dbAdapter.cleanupStaleEvents(Table.EVENTS);
            dbAdapter.cleanupStaleEvents(Table.PROFILE_EVENTS);
            dbAdapter.cleanupStaleEvents(Table.PUSH_NOTIFICATION_VIEWED);
            dbAdapter.cleanUpPushNotifications();
        }
        return dbAdapter;
    }

    /**
     * Only call async
     */
    @Override
    public void clearQueues(final Context context) {
        synchronized (ctLockManager.getEventLock()) {

            DBAdapter adapter = loadDBAdapter(context);
            Table tableName = Table.EVENTS;

            adapter.removeEvents(tableName);
            tableName = Table.PROFILE_EVENTS;
            adapter.removeEvents(tableName);

            clearUserContext(context);
        }
    }

    //Session
    private void clearIJ(Context context) {
        final SharedPreferences prefs = StorageHelper.getPreferences(context, Constants.NAMESPACE_IJ);
        final SharedPreferences.Editor editor = prefs.edit();
        editor.clear();
        StorageHelper.persist(editor);
    }

    //Session
    private void clearLastRequestTimestamp(Context context) {
        StorageHelper.putInt(context, StorageHelper.storageKeyWithSuffix(config, Constants.KEY_LAST_TS), 0);
    }

    //Session
    private void clearUserContext(final Context context) {
        clearIJ(context);
        clearFirstRequestTimestampIfNeeded(context);
        clearLastRequestTimestamp(context);
    }

    //Session
    private void clearFirstRequestTimestampIfNeeded(Context context) {
        StorageHelper.putInt(context, StorageHelper.storageKeyWithSuffix(config, Constants.KEY_FIRST_TS), 0);
    }

    // helper extracts the cursor data from the db object

    @Override
    QueueCursor getPushNotificationViewedQueuedEvents(final Context context, final int batchSize,
            final QueueCursor previousCursor) {
        return getQueueCursor(context, Table.PUSH_NOTIFICATION_VIEWED, batchSize, previousCursor);
    }

    @Override
    QueueCursor getQueueCursor(final Context context, final Table table, final int batchSize,
            final QueueCursor previousCursor) {
        synchronized (ctLockManager.getEventLock()) {
            DBAdapter adapter = loadDBAdapter(context);
            Table tableName = (previousCursor != null) ? previousCursor.getTableName() : table;

            // if previousCursor that means the batch represented by the previous cursor was processed so remove those from the db
            if (previousCursor != null) {
                adapter.cleanupEventsFromLastId(previousCursor.getLastId(), previousCursor.getTableName());
            }

            // grab the new batch
            QueueCursor newCursor = new QueueCursor();
            newCursor.setTableName(tableName);
            JSONObject queuedDBEvents = adapter.fetchEvents(tableName, batchSize);
            newCursor = updateCursorForDBObject(queuedDBEvents, newCursor);

            return newCursor;
        }
    }

    // note : only works with  Queue Cursor of Events table. for other cursors, will override its data with Event table's data
    @Override
    QueueCursor getQueuedDBEvents(final Context context, final int batchSize, final QueueCursor previousCursor) {

        synchronized (ctLockManager.getEventLock()) {
            QueueCursor newCursor = getQueueCursor(context, Table.EVENTS, batchSize, previousCursor);

            if (newCursor.isEmpty() && newCursor.getTableName().equals(Table.EVENTS)) {
                newCursor = getQueueCursor(context, Table.PROFILE_EVENTS, batchSize, null);
            }

            return newCursor.isEmpty() ? null : newCursor;
        }
    }

    @SuppressWarnings("SameParameterValue")
    public QueueCursor getQueuedEvents(final Context context, final int batchSize, final QueueCursor previousCursor,
            final EventGroup eventGroup) {
        if (eventGroup == EventGroup.PUSH_NOTIFICATION_VIEWED) {
            config.getLogger().verbose(config.getAccountId(), "Returning Queued Notification Viewed events");
            return getPushNotificationViewedQueuedEvents(context, batchSize, previousCursor);
        } else {
            config.getLogger().verbose(config.getAccountId(), "Returning Queued events");
            return getQueuedDBEvents(context, batchSize, previousCursor);
        }
    }

    //Event
    @WorkerThread
    @Override
    public void queueEventToDB(final Context context, final JSONObject event, final int type) {
        Table table = (type == Constants.PROFILE_EVENT) ? Table.PROFILE_EVENTS
                : Table.EVENTS;
        queueEventInternal(context, event, table);
    }

    @WorkerThread
    @Override
    public void queuePushNotificationViewedEventToDB(final Context context, final JSONObject event) {
        queueEventInternal(context, event, Table.PUSH_NOTIFICATION_VIEWED);
    }

    /* calling this function Will set items from jsonObject to queue cursor.
     * if jsonObject is null, will return the queue cursor as it is
     * else
     *  - it will try taking the first key and set it at as cursor's Last Id
     *  - it will try taking the value of first key, cast it as jsonArray and set it as cursor's data
     *
     * therefore, the json must be of format : {string : jsonArray }
     * */
    @Override
    QueueCursor updateCursorForDBObject(final JSONObject dbObject, final QueueCursor cursor) {
        if (dbObject == null) {
            return cursor;
        }

        Iterator<String> keys = dbObject.keys();
        if (keys.hasNext()) {
            String key = keys.next();
            cursor.setLastId(key);
            try {
                cursor.setData(dbObject.getJSONArray(key));
            } catch (JSONException e) {
                cursor.setLastId(null);
                cursor.setData(null);
            }
        }

        return cursor;
    }

    @WorkerThread
    private void queueEventInternal(final Context context, final JSONObject event, Table table) {
        synchronized (ctLockManager.getEventLock()) {
            DBAdapter adapter = loadDBAdapter(context);
            long returnCode = adapter.storeObject(event, table);

            if (returnCode > 0) {
                config.getLogger().debug(config.getAccountId(), "Queued event: " + event.toString());
                config.getLogger()
                        .verbose(config.getAccountId(),
                                "Queued event to DB table " + table + ": " + event.toString());
            }
        }
    }
}

File: clevertap-core/src/main/java/com/clevertap/android/sdk/db/DBManager.kt
package com.clevertap.android.sdk.db

import android.content.Context
import androidx.annotation.WorkerThread
import com.clevertap.android.sdk.CTLockManager
import com.clevertap.android.sdk.Constants
import com.clevertap.android.sdk.ILogger
import com.clevertap.android.sdk.db.Table.EVENTS
import com.clevertap.android.sdk.db.Table.PROFILE_EVENTS
import com.clevertap.android.sdk.db.Table.PUSH_NOTIFICATION_VIEWED
import com.clevertap.android.sdk.events.EventGroup
import org.json.JSONObject

internal class DBManager constructor(
    private val accountId: String,
    private val logger: ILogger,
    private val databaseName: String,
    private val ctLockManager: CTLockManager,
    private val dbEncryptionHandler: DBEncryptionHandler,
) : BaseDatabaseManager {

    private companion object {
        private const val USER_EVENT_LOG_ROWS_PER_USER = 2_048 + 256 // events + profile props
        private const val USER_EVENT_LOG_ROWS_THRESHOLD = 5 * USER_EVENT_LOG_ROWS_PER_USER
    }

    private var dbAdapter: DBAdapter? = null

    @WorkerThread
    @Synchronized
    override fun loadDBAdapter(context: Context): DBAdapter {
        var dbAdapter = this.dbAdapter
        if (dbAdapter == null) {
            dbAdapter = DBAdapter(
                context = context,
                databaseName = databaseName,
                accountId = accountId,
                logger = logger,
                dbEncryptionHandler = dbEncryptionHandler
            )
            this.dbAdapter = dbAdapter
            dbAdapter.cleanupStaleEvents(EVENTS)
            dbAdapter.cleanupStaleEvents(PROFILE_EVENTS)
            dbAdapter.cleanupStaleEvents(PUSH_NOTIFICATION_VIEWED)
            dbAdapter.cleanUpPushNotifications()
            dbAdapter.userEventLogDAO()
                .cleanUpExtraEvents(USER_EVENT_LOG_ROWS_THRESHOLD, USER_EVENT_LOG_ROWS_PER_USER)
        }
        return dbAdapter
    }

    /**
     * Deletes all the events from the event and profile event queues.
     *
     * @param context The Android context.
     */
    @WorkerThread
    override fun clearQueues(context: Context) {
        val adapter = loadDBAdapter(context)
        var tableName = EVENTS
        adapter.removeEvents(tableName)
        tableName = PROFILE_EVENTS
        adapter.removeEvents(tableName)
    }

    /**
     * Main entry point for getting queued events
     * Now returns a combined batch of events and profile events with cleanup info
     */
    override fun getQueuedEvents(
        context: Context,
        batchSize: Int,
        eventGroup: EventGroup
    ): QueueData {
        return when (eventGroup) {
            EventGroup.PUSH_NOTIFICATION_VIEWED -> {
                logger.verbose(accountId, "Returning Queued Notification Viewed events")
                getPushNotificationViewedQueuedEvents(context, batchSize)
            }
            else -> {
                logger.verbose(accountId, "Returning combined queued events")
                getCombinedQueuedEvents(context, batchSize)
            }
        }
    }

    /**
     * Fetches a combined batch of events from both events and profileEvents tables
     * Returns QueueData with events data and ids, also if there are more events to fetch
     */
    override fun getCombinedQueuedEvents(context: Context, batchSize: Int): QueueData {
        synchronized(ctLockManager.eventLock) {
            val adapter = loadDBAdapter(context)

            // Fetch combined batch of events with cleanup info
            return adapter.fetchCombinedEvents(batchSize)
        }
    }

    /**
     * Cleans up successfully sent events
     * Should be called from NetworkManager after successful transmission
     *
     * @param context Android context
     * @param eventIds List of event IDs to clean up from events table
     * @param profileEventIds List of event IDs to clean up from profileEvents table
     */
    override fun cleanupSentEvents(
        context: Context,
        eventIds: List<String>,
        profileEventIds: List<String>
    ): Boolean {
        synchronized(ctLockManager.eventLock) {
            // Return true if nothing to clean up
            if (eventIds.isEmpty() && profileEventIds.isEmpty()) {
                return true
            }

            try {
                val adapter = loadDBAdapter(context)

                // Clean up events from events table
                if (eventIds.isNotEmpty()) {
                    //adapter.cleanupEventsByIds(EVENTS, eventIds)
                    adapter.cleanupEventsFromLastId(eventIds[eventIds.size-1], EVENTS)
                    logger.verbose(
                        accountId,
                        "Cleaned ${eventIds.size} events from events table"
                    )
                }

                // Clean up events from profileEvents table
                if (profileEventIds.isNotEmpty()) {
                    //adapter.cleanupEventsByIds(PROFILE_EVENTS, profileEventIds)
                    adapter.cleanupEventsFromLastId(profileEventIds[profileEventIds.size-1], PROFILE_EVENTS)
                    logger.verbose(
                        accountId,
                        "Cleaned ${profileEventIds.size} events from profileEvents table"
                    )
                }

                return true

            } catch (e: Exception) {
                logger.verbose(
                    accountId,
                    "Error during cleanup of sent events",
                    e
                )
                return false
            }
        }
    }

    override fun cleanupPushNotificationEvents(context: Context, ids: List<String>) : Boolean {
        synchronized(ctLockManager.eventLock) {
            // Return true if nothing to clean up
            if (ids.isEmpty()) {
                return true
            }

            try {
                val adapter = loadDBAdapter(context)
                // Clean up events from profileEvents table
                if (ids.isNotEmpty()) {
                    //adapter.cleanupEventsByIds(PUSH_NOTIFICATION_VIEWED, ids)
                    adapter.cleanupEventsFromLastId(ids[ids.size - 1], PUSH_NOTIFICATION_VIEWED)
                    logger.verbose(
                        accountId,
                        "Cleaned ${ids.size} events from Push impressions table"
                    )
                }
                return true
            } catch (e: Exception) {
                logger.verbose(
                    accountId,
                    "Error during cleanup of notification sent events",
                    e
                )
                return false
            }
        }
    }

    /**
     * Handles push notification viewed events separately
     * These remain in their own queue
     */
    override fun getPushNotificationViewedQueuedEvents(
        context: Context,
        batchSize: Int
    ): QueueData {
        synchronized(ctLockManager.eventLock) {
            val adapter = loadDBAdapter(context)

            // Use the optimized fetchEvents method that returns QueueData
            return adapter.fetchEvents(PUSH_NOTIFICATION_VIEWED, batchSize)
        }
    }

    //Event
    @WorkerThread
    override fun queueEventToDB(context: Context, event: JSONObject, type: Int) {
        val table = if (type == Constants.PROFILE_EVENT) PROFILE_EVENTS else EVENTS
        queueEventForTable(context, event, table)
    }

    @WorkerThread
    override fun queuePushNotificationViewedEventToDB(context: Context, event: JSONObject) {
        queueEventForTable(context, event, PUSH_NOTIFICATION_VIEWED)
    }

    @WorkerThread
    private fun queueEventForTable(context: Context, event: JSONObject, table: Table) {
        synchronized(ctLockManager.eventLock) {
            val adapter = loadDBAdapter(context)
            val returnCode = adapter.storeObject(event, table)
            if (returnCode > 0) {
                logger.debug(accountId, "Queued event: $event")
                logger.verbose(accountId, "Queued event to DB table $table: $event")
            }
        }
    }
}
